src/modules/flow/persistence/efivarfs-storage.c

```diff
@@ -0,0 +1,167 @@
+/*
+ * This file is part of the Soletta Project
+ *
+ * Copyright (C) 2015 Intel Corporation. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in
+ *     the documentation and/or other materials provided with the
+ *     distribution.
+ *   * Neither the name of Intel Corporation nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "efivarfs-storage.h"
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "sol-buffer.h"
+#include "sol-log.h"
+#include "sol-util.h"
+#include "sol-util-file.h"
+
+#define EFIVARS_DEFAULT_ATTR 0x7
+#define SOLETTA_EFIVARS_GUID "076027a8-c791-41d7-940f-3d465869f821"
+#define EFIVARFS_VAR_PATH "/sys/firmware/efi/efivars/%s-" SOLETTA_EFIVARS_GUID
+
+struct efi_var {
+    uint32_t attributes;
+    uint8_t data[];
+};
+
+int
+efivars_write(const char *name, void *data, size_t size)
+{
+#ifdef FEATURE_FILESYSTEM
+    struct efi_var *efi_var;
+    FILE *file;
+    char path[PATH_MAX];
+    int r;
+
+    r = snprintf(path, sizeof(path), EFIVARFS_VAR_PATH, name);
+    if (r < 0 || r >= PATH_MAX) {
+        SOL_WRN("Could not create path for efivars persistence file [%s]", path);
+        return -EINVAL;
+    }
+
+    efi_var = calloc(1, sizeof(uint32_t) + size);
+    SOL_NULL_CHECK(efi_var, -ENOMEM);
+
+    efi_var->attributes = EFIVARS_DEFAULT_ATTR;
+    memcpy(efi_var->data, data, size);
```

Comment:
do we need to write as a single call? Otherwise we could skip allocating `efi_var` only to do that.

Also, it is confusing to see `sizeof(uint32_t)` in there, it's better to write `sizeof(struct efi_var)`, given your declaration it should be `sizeof(uint32_t)`.